Zotero/Scenes/Detail/HTML:EPUB/ViewModels/HtmlEpubReaderActionHandler.swift

```diff
@@ -0,0 +1,876 @@
+            let jsArrayData = try JSONSerialization.data(withJSONObject: [UInt8](data))
+            guard let jsArrayString = String(data: jsArrayData, encoding: .utf8) else {
+                DDLogError("HtmlEpubReaderActionHandler: can't convert data to string")
+                return
+            }
+            let (sortedKeys, annotations, json, token, rawPage) = loadAnnotationsAndJson(in: viewModel)
+            let type: String
+            let page: HtmlEpubReaderState.DocumentData.Page?
+
+            switch viewModel.state.url.pathExtension.lowercased() {
+            case "epub":
+                type = "epub"
+                page = .epub(cfi: rawPage)
+
+            case "html", "htm":
+                type = "snapshot"
+                if let scrollYPercent = Double(rawPage) {
+                    page = .html(scrollYPercent: scrollYPercent)
+                } else {
+                    DDLogError("HtmlEPubReaderActionHandler: incompatible lastIndexPage stored for \(viewModel.state.key) - \(rawPage)")
+                    page = nil
+                }
+
+            default:
+                throw HtmlEpubReaderState.Error.incompatibleDocument
+            }
+
+            let documentData = HtmlEpubReaderState.DocumentData(type: type, buffer: jsArrayString, annotationsJson: json, page: page)
+            update(viewModel: viewModel) { state in
+                state.sortedKeys = sortedKeys
+                state.annotations = annotations
+                state.documentData = documentData
+                state.notificationToken = token
+                state.changes = .annotations
+            }
+        } catch let error {
+            DDLogError("HtmlEpubReaderActionHandler: could not load document - \(error)")
+        }
+    }
+
+    private func loadAnnotationsAndJson(in viewModel: ViewModel<HtmlEpubReaderActionHandler>) -> ([String], [String: HtmlEpubAnnotation], String, NotificationToken?, String) {
+        do {
+            let pageIndexRequest = ReadDocumentDataDbRequest(attachmentKey: viewModel.state.key, libraryId: viewModel.state.library.identifier)
+            let pageIndex = try dbStorage.perform(request: pageIndexRequest, on: .main)
+            let annotationsRequest = ReadAnnotationsDbRequest(attachmentKey: viewModel.state.key, libraryId: viewModel.state.library.identifier)
+            let items = try dbStorage.perform(request: annotationsRequest, on: .main)
+            var sortedKeys: [String] = []
+            var annotations: [String: HtmlEpubAnnotation] = [:]
+            var jsons: [[String: Any]] = []
+
+            for item in items {
+                guard let (annotation, json) = item.htmlEpubAnnotation else { continue }
+                jsons.append(json)
+                sortedKeys.append(annotation.key)
+                annotations[item.key] = annotation
+            }
+
+            let jsonString = WebViewEncoder.encodeAsJSONForJavascript(jsons)
```

Comment:
This should not be encoded in base 64.
E.g.:
```suggestion
            let jsonString = WebViewEncoder.encodeAsJSONForJavascript(jsons)
```
Better yet if `WebViewEncoder.encodeAsJSONForJavascript` accepts an extra parameter that determines the final encoding. I guess that this method should return `String?` instead of `String`, so the caller determine what the value should be in case it fails, e.g. here it is `[]`.